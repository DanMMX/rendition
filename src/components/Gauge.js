import * as React from 'react'
const Pie = require('paths-js/pie')
import styled, { withTheme } from 'styled-components'
const map = require('lodash/map')
import { compose } from 'recompose'
import { px } from '../utils'

import asRendition from '../asRendition'
import theme from '../theme'

const { font } = theme

const StatusGaugeWrapper = styled.div`
  display: block;
  color: ${({ theme }) => theme.colors.text.main};

  /* fix edge hover missplacing bug
    on svg text inside anchors
  */
  &,
  &:hover {
    text-decoration: none;

    text {
      text-decoration: none !important;
    }

    .gauge__graph__total-label {
      transform: translateX(0);
    }
  }

  .gauge__graph {
    display: block;
    width: 100%;
    max-width: 180px;
    margin: 0 auto 10px;
  }

  .gauge__graph__total-label {
    font-family: ${font};
    font-size: 5px;
    fill: #b8b8b8;
    text-anchor: middle;
  }

  .gauge__legend {
    list-style: none;
    padding: 5px;
    margin: 0 auto;
    background: white;
    border-radius: 5px;
    max-width: 240px;
    border: 1px solid #ececec;
    overflow: hidden;
  }

  .gauge__legend-item {
    display: flex;
    padding: 3px 0 3px 2px;
    line-height: 9px;
    width: 50%;
    float: left;
  }

  .gauge__legend-color {
    display: inline-block;
    flex: 0 0 auto;
    width: 10px;
    height: 10px;
    margin-right: 4px;
  }

  .gauge__legend-name {
    font-size: 11px;
  }
`

const GaugeTotalText = styled.text`
  font-family: ${font};
  font-size: ${props => px(props.count > 999 ? 20 : 26)};
  fill: #5e5e5e;
  letter-spacing: -1px;
  text-anchor: middle;
  transform: translateX(0);
`

const GaugeTotal = ({ count, ...props }) => (
  <GaugeTotalText count={count} {...props}>
    {count}
  </GaugeTotalText>
)

/**
 * data: [{
 *   key: string
 *   value: number
 *   color: string
 *   name: string
 * }]
 */
const StatusGauge = ({
  placeholderColor = theme.colors.gray.light,
  legend,
  data,
  title,
  ...props
}) => {
  const count = data.reduce((carry, item) => {
    return carry + item.value
  }, 0)

  const graph = Pie({
    accessor: item => item.value,
    center: [0, 0],
    data: data,
    R: 46,
    r: 25
  })

  const gauge = graph.curves
  return (
    <StatusGaugeWrapper {...props}>
      <svg
        className='gauge__graph'
        viewBox='0 0 100 100'
        xmlns='http://www.w3.org/2000/svg'
      >
        <circle
          fill=' white'
          stroke='#dadada'
          strokeWidth='5'
          cx='50'
          cy='50'
          r='46'
        />

        <g>
          <circle fill={placeholderColor} cx='50' cy='50' r='46' />
          <circle fill='white' cx='50' cy='50' r='25' />
        </g>

        <g transform='translate(50, 50)'>
          {map(gauge, segment => {
            if (!count) {
              return
            }
            if (segment.item.value === count) {
              return (
                <g key={segment.item.key || segment.item.name}>
                  <path
                    transform='translate(-100, -100)'
                    fill={segment.item.color}
                    d='M100 100m-46,0a46,46,0 1,0 92,0a 46,46 0 1,0 -92,0zM100 100m-25,0a25,25,0 0,1 50,0a 25,25 0 0,1 -50,0z'
                  />
                </g>
              )
            }
            return (
              <g key={segment.item.key || segment.item.name}>
                <path
                  fill={segment.item.color}
                  d={segment.sector.path.print()}
                />
              </g>
            )
          })}
        </g>
        <text className='gauge__graph__total-label' x='50' y='41'>
          {title ? ` ${title.toUpperCase()}` : 'TOTAL'}
        </text>
        <GaugeTotal x='50' y='63' count={count} />
      </svg>

      <ul className='gauge__legend'>
        {map(data, item => {
          return (
            <li key={item.key || item.name} className={`gauge__legend-item`}>
              <span
                className={`gauge__legend-color`}
                style={{ backgroundColor: item.color }}
              >
                &nbsp;
              </span>
              <span className='gauge__legend-name'>{item.name}</span>
            </li>
          )
        })}
      </ul>
    </StatusGaugeWrapper>
  )
}

export default compose(withTheme, asRendition)(StatusGauge)
