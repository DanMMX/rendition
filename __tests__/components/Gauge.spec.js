/* globals expect, test */
import React from 'react'
import renderer from 'react-test-renderer'
import Provider from '../../src/components/Provider'
import Gauge from '../../src/components/Gauge'

const data = [
  {
    value: 8,
    name: 'Vanilla',
    color: 'beige'
  },
  {
    value: 9,
    name: 'Chocolate',
    color: 'brown'
  },
  {
    value: 7,
    name: 'Strawberry',
    color: 'pink'
  }
]

test('Gauge renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Gauge title='Ice Cream' data={data} />
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
