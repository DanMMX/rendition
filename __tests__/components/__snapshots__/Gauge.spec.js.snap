// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gauge renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <div
    className="sc-bxivhb ljpeSw sc-bwzfXH hBrfBJ"
  >
    <svg
      className="gauge__graph"
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="50"
        cy="50"
        fill=" white"
        r="46"
        stroke="#dadada"
        strokeWidth="5"
      />
      <g>
        <circle
          cx="50"
          cy="50"
          fill="#f4f4f4"
          r="46"
        />
        <circle
          cx="50"
          cy="50"
          fill="white"
          r="25"
        />
      </g>
      <g
        transform="translate(50, 50)"
      >
        <g>
          <path
            d="M 0 -46 A 46 46 0 0 1 39.837169 23 L 21.650635 12.5 A 25 25 0 0 0 0 -25 Z "
            fill="beige"
          />
        </g>
        <g>
          <path
            d="M 39.837169 23 A 46 46 0 0 1 -44.432588 11.905676 L -24.148146 6.470476 A 25 25 0 0 0 21.650635 12.5 Z "
            fill="brown"
          />
        </g>
        <g>
          <path
            d="M -44.432588 11.905676 A 46 46 0 0 1 -0 -46 L -0 -25 A 25 25 0 0 0 -24.148146 6.470476 Z "
            fill="pink"
          />
        </g>
      </g>
      <text
        className="gauge__graph__total-label"
        x="50"
        y="41"
      >
         ICE CREAM
      </text>
      <text
        className="sc-htpNat gwYnzy"
        x="50"
        y="63"
      >
        24
      </text>
    </svg>
    <ul
      className="gauge__legend"
    >
      <li
        className="gauge__legend-item"
      >
        <span
          className="gauge__legend-color"
          style={
            Object {
              "backgroundColor": "beige",
            }
          }
        >
           
        </span>
        <span
          className="gauge__legend-name"
        >
          Vanilla
        </span>
      </li>
      <li
        className="gauge__legend-item"
      >
        <span
          className="gauge__legend-color"
          style={
            Object {
              "backgroundColor": "brown",
            }
          }
        >
           
        </span>
        <span
          className="gauge__legend-name"
        >
          Chocolate
        </span>
      </li>
      <li
        className="gauge__legend-item"
      >
        <span
          className="gauge__legend-color"
          style={
            Object {
              "backgroundColor": "pink",
            }
          }
        >
           
        </span>
        <span
          className="gauge__legend-name"
        >
          Strawberry
        </span>
      </li>
    </ul>
  </div>
</div>
`;
